pkg/task/task.go

```diff
@@ -0,0 +1,68 @@
+package task
+
+import "time"
+
+// TaskState (kind: int) represents the last known state of a task.
+// A task can be in one of three states
+// TaskStateScheduled: this is the initial state of the task when it enters the queue.
+// TaskStateProcessing: once work begins on the task, it is put into this state.
+// TaskStateComplete: work is no longer being done on this task. client should check task result.
+type TaskState int
+
+const (
+	StateScheduled TaskState = iota
+	StateProcessing
+	StateComplete
+)
+
+func (t TaskState) String() string {
+	return [...]string{
+		"StateRequested",
+		"StateProcessing",
+		"StateComplete",
+	}[t]
+}
+
+// TaskResultStatus (kind: int) is a status code for completed tasks.
+// TaskResultNone: initial status, No status, probably becasue the task is incomplete.
+// TaskResultSuccess: the task has completed without an error.
+// TaskResultFail: the task has completed with a failure.
+// TaskResultAbort: testground encountered an error and the task has not been scheduled.
+// testground will respond to client with a TaskResultAbort status if it is unable to enqueue
+// the reqeusted task.
+type TaskResultStatus int
+
+const (
+	ResultNone TaskResultStatus = iota
+	ResultSuccess
+	ResultFail
+	ResultAbort
+)
+
+func (t TaskResultStatus) String() string {
+	return [...]string{
+		"ResultNone",
+		"ResultSuccess",
+		"ResultFail",
+		"ResultAbort",
+	}[t]
+}
+
+// TaskResult (kind: struct)  contains a status code. If the status is not TaskResultNone or
+// TaskResultSuccess, relevant errors will be included in this struct.
+type TaskResult struct {
+	Status TaskResultStatus `json:"status"`
+	Errors []error          `json:"errors"`
+}
+
+// Task (kind: struct) contains metadata about a testground task. This schema is used to store
+// metadata in our task storage database as well as the wire format returned when clients get the
+// state of a running or secheduled task.
+type Task struct {
+	Version  int        `json:"version"`  // Schema version
+	Priority int        `json:"priority"` // scheduling priority
+	Created  time.Time  `json:"created"`  // datetime created
+	ID       string     `json:"id"`       // unique identifier for this task
+	State    TaskState  `json:"state"`    // State of the task
+	Result   TaskResult `json:"result"`   // result of the task, when terminal.
```

Comment:
We currently don't have a good definition of what the result of a task is:
1. Is it the result from the pods/containers execution vs `outcome` in test plan.
2. Is one failed test plan instance considered a task failure?

I think it makes sense dropping `Result` altogether from the `Task` struct, or modelling it to give information per test instance based on the pod/container status (9999 completed, 1 failed), or something like that.

Reply:
We definitely need to figure out what we want to include in the result struct.